upgrade/util.go

```diff
@@ -43,6 +43,13 @@ type Context struct {
 	//
 	UpstreamProviderName string
 
+	// The desired version of pulumi/{pkg,sdk} to link to.
+	//
+	// If TargetPulumiVersion is nil, then pulumi/{pkg,sdk} should follow the bridge.
+	//
+	// Otherwise, we will `require` the described version.
+	TargetPulumiVersion string
```

Comment:
Could be a Ref.

Reply:
It can't be `<latest>` and we never do anything  but stringify it.

Reply:
Sure, no strong opinions. It would let you distinguish if it's a hash or not easily.

Reply:
Changed to use `Ref`. I don't think we need to do anything special, but I agree it's the correct long term value. May as well take it now.